pkg/collector/corechecks/snmp/discovery/discovery.go

```diff
@@ -0,0 +1,294 @@
+		go d.runWorker(w, jobs)
+	}
+
+	discoveryTicker := time.NewTicker(time.Duration(d.config.DiscoveryInterval) * time.Second)
+
+	for {
+		log.Debugf("subnet %s: Run discovery", d.config.Network)
+		startingIP := make(net.IP, len(subnet.startingIP))
+		copy(startingIP, subnet.startingIP)
+		for currentIP := startingIP; subnet.network.Contains(currentIP); incrementIP(currentIP) {
+
+			if ignored := subnet.config.IsIPIgnored(currentIP); ignored {
+				continue
+			}
+
+			jobIP := make(net.IP, len(currentIP))
+			copy(jobIP, currentIP)
+			job := snmpJob{
+				subnet:    &subnet,
+				currentIP: jobIP,
+			}
+			jobs <- job
+
+			select {
+			case <-d.stop:
+				log.Debugf("subnet %s: Stop scheduling devices", d.config.Network)
+				return
+			default:
+			}
+		}
+
+		select {
+		case <-d.stop:
+			log.Debugf("subnet %s: Stop scheduling devices", d.config.Network)
+			return
+		case <-discoveryTicker.C:
+		}
+	}
+}
+
+func (d *Discovery) checkDevice(job snmpJob) error {
+	deviceIP := job.currentIP.String()
+	config := *job.subnet.config // shallow copy
+	config.IPAddress = deviceIP
+	sess, err := session.NewSession(&config)
+	if err != nil {
+		return fmt.Errorf("error configure session for ip %s: %v", deviceIP, err)
+	}
+	entityID := job.subnet.config.DiscoveryDigest(deviceIP)
+	if err := sess.Connect(); err != nil {
+		log.Debugf("subnet %s: SNMP connect to %s error: %v", d.config.Network, deviceIP, err)
+		d.deleteDevice(entityID, job.subnet)
+	} else {
+		defer sess.Close()
+
+		oids := []string{sysObjectIDOid}
+		// Since `params<GoSNMP>.ContextEngineID` is empty
+		// `params.Get` might lead to multiple SNMP GET calls when using SNMP v3
+		value, err := sess.Get(oids)
+		if err != nil {
+			log.Debugf("subnet %s: SNMP get to %s error: %v", d.config.Network, deviceIP, err)
+			d.deleteDevice(entityID, job.subnet)
+		} else if len(value.Variables) < 1 || value.Variables[0].Value == nil {
+			log.Debugf("subnet %s: SNMP get to %s no data", d.config.Network, deviceIP)
+			d.deleteDevice(entityID, job.subnet)
+		} else {
+			log.Debugf("subnet %s: SNMP get to %s success: %v", d.config.Network, deviceIP, value.Variables[0].Value)
+			d.createDevice(entityID, job.subnet, deviceIP, true)
+		}
+	}
+	return nil
+}
+
+func (d *Discovery) createDevice(entityID string, subnet *snmpSubnet, deviceIP string, writeCache bool) {
+	deviceCk, err := devicecheck.NewDeviceCheck(subnet.config, deviceIP)
+	if err != nil {
+		// should not happen since the deviceCheck is expected to be valid at this point
+		// and are only changing the device ip
+		log.Warnf("subnet %s: failed to create new device check `%s`: %s", d.config.Network, deviceIP, err)
+		return
+	}
+
+	d.discDevMu.Lock()
+	defer d.discDevMu.Unlock()
+
+	if _, present := d.discoveredDevices[entityID]; present {
+		return
+	}
+	device := Device{
+		entityID:    entityID,
+		deviceIP:    deviceIP,
+		deviceCheck: deviceCk,
+	}
+	d.discoveredDevices[entityID] = device
+	subnet.devices[entityID] = deviceIP
```

Comment:
subnet deserves a mutex too

Reply:
We are already protected by this mutex https://github.com/DataDog/datadog-agent/blob/0295e8e7ab29a2d7763b856d79e02a39122c4aec/pkg/collector/corechecks/snmp/discovery/discovery.go#L210-L211

I added some comments here 004f765fc , let me know if that's ok for you. 

Since everything is internal to discovery.go, I think this should be fine for now to rely on a single mutex.